// office-ui-fabric-react/lib-es2015/components/Rating/Rating.js
import * as tslib_1 from "tslib";
import * as React from 'react';
import { BaseComponent, css, getId } from '../../Utilities';
import { Icon } from '../../Icon';
import { FocusZone, FocusZoneDirection } from '../../FocusZone';
import { RatingSize } from './Rating.types';
import * as stylesImport from './Rating.scss';
import { format } from '../../Utilities';
var styles = stylesImport;
var RatingStar = function (props) {
    return (React.createElement("div", { className: css('ms-RatingStar-container', styles.ratingStar), key: props.id },
        React.createElement(Icon, { className: css('ms-RatingStar-back', styles.ratingStarBack, (_a = {},
                _a['is-disabled ' + styles.ratingStarDisabled] = props.disabled,
                _a)), iconName: 'FavoriteStarFill' }),
        !props.disabled &&
            React.createElement(Icon, { className: css('ms-RatingStar-front', styles.ratingStarFront), iconName: 'FavoriteStarFill', style: { width: props.fillPercentage + '%' } })));
    var _a;
};
var Rating = /** @class */ (function (_super) {
    tslib_1.__extends(Rating, _super);
    function Rating(props) {
        var _this = _super.call(this, props) || this;
        _this.state = {
            rating: _this._getInitialValue(props)
        };
        _this._id = getId('Rating');
        _this._labelId = getId('RatingLabel');
        return _this;
    }
    Rating.prototype.componentWillReceiveProps = function (nextProps) {
        if (typeof nextProps.rating !== 'undefined' && nextProps.rating !== this.state.rating) {
            this.setState({
                rating: this._getClampedRating(nextProps.rating)
            });
        }
    };
    Rating.prototype.render = function () {
        var stars = [];
        for (var i = this.props.min; i <= this.props.max; i++) {
            var ratingStarProps = {
                fillPercentage: this._getFillingPercentage(i),
                disabled: this.props.disabled ? true : false
            };
            stars.push(React.createElement("button", tslib_1.__assign({ className: css('ms-Rating-button', styles.ratingButton, (_a = {},
                    _a['ms-Rating--large ' + styles.rootIsLarge] = this.props.size === RatingSize.Large,
                    _a['ms-Rating--small ' + styles.rootIsSmall] = this.props.size !== RatingSize.Large,
                    _a)), key: i }, ((i === Math.ceil(this.state.rating)) ? { 'data-is-current': true } : {}), { onFocus: this._onFocus.bind(this, i), disabled: this.props.disabled || this.props.readOnly ? true : false, role: 'presentation' }),
                this._getLabel(i),
                React.createElement(RatingStar, tslib_1.__assign({ key: i + 'rating' }, ratingStarProps))));
        }
        return (React.createElement("div", { className: css('ms-Rating-star'), "aria-label": this.props.getAriaLabel ? this.props.getAriaLabel(this.state.rating ? this.state.rating : 0, this.props.max) : '' },
            React.createElement(FocusZone, { direction: FocusZoneDirection.horizontal, tabIndex: this.props.readOnly ? 0 : -1, className: css('ms-Rating-focuszone', styles.ratingFocusZone) }, stars)));
        var _a;
    };
    Rating.prototype._onFocus = function (value, ev) {
        this.setState({
            rating: value
        });
        var onChanged = this.props.onChanged;
        if (onChanged) {
            onChanged(value);
        }
    };
    Rating.prototype._getLabel = function (rating) {
        var text = this.props.ariaLabelFormat || '';
        return (React.createElement("span", { id: this._labelId + "-" + rating, className: css('ms-Rating-labelText', styles.labelText) }, format(text, rating, this.props.max)));
    };
    Rating.prototype._getInitialValue = function (props) {
        if (typeof props.rating === 'undefined') {
            return props.min;
        }
        if (props.rating === null) {
            return null;
        }
        return this._getClampedRating(props.rating);
    };
    Rating.prototype._getClampedRating = function (rating) {
        return Math.min(Math.max(rating, this.props.min), this.props.max);
    };
    Rating.prototype._getFillingPercentage = function (starPosition) {
        var ceilValue = Math.ceil(this.state.rating);
        var fillPercentage = 100;
        if (starPosition === this.state.rating) {
            fillPercentage = 100;
        }
        else if (starPosition === ceilValue) {
            fillPercentage = 100 * (this.state.rating % 1);
        }
        else if (starPosition > ceilValue) {
            fillPercentage = 0;
        }
        return fillPercentage;
    };
    Rating.defaultProps = {
        min: 1,
        max: 5
    };
    return Rating;
}(BaseComponent));
export { Rating };
